### DISCUSSION TITLE
Why is it append instead of replace when adding fields

### DISCUSSION THREAD

**Discussion by 3tbraden** [uber-go/zap#1361] [Q&A] [answered]
I was trying to have some common fields in all logs and I used `.With()` and `.WithOptions` for that. The problem is, when I pass the logger to a child function, which logs the same field again, the value in the field is repeated.

i.e.
```Go
func fnA() {
  logger = logger.With(zap.String("s", "sth"))
  ctx = NewContext(ctx, logger)
  fnB(ctx)
}

func fnB(ctx) {
  logger = FromContext(ctx)
  logger.Info("s", "else")
}
```
the log from `fnB` above will become `{"s": "sthelse"}`. Is it possible to replace instead of append when adding the fields? Any reason why it should be append?

**Comment by abhinav** [comment, marked as answer]
Hey @3tbraden. Yeah, this is a fair question.
Part of the reason for the choice is this: it's difficult to get the performance we want out of Zap by doing this with replace.
Particularly, note that `With` encodes keys at the time `logger.With` is called.
Going back and changing what's already been encoded to fix a duplicate would require decoding and re-encoding that blob.

There's also been some prior discussion about this in #622.
